### PR TITLE
[System-Probe] Build Probe on Windows

### PR DESCRIPTION
**Pull request by stb2132**
### What does this PR do?

This PR gets the system probe building on windows. The program can run, but only accepts connections on the /connections/ endpoint, and returns a dummy value. 
### Motivation

Begin getting feedback

### Additional Notes

Anything else we should know when reviewing?

### CODE REVIEW

File: cmd/system-probe/main_common.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+	Version   string
+	GitCommit string
+	GitBranch string
+	BuildDate string
+)
+
+const loggerName = ddconfig.LoggerName("SYS-PROBE")
+
+func runAgent(exit chan bool) {
+
```
Comment (sunhay) [inline cmd/system-probe/main_common.go]: nit: remove empty space

File: cmd/system-probe/main.go [hunk 2]
```diff
@@ -54,108 +30,29 @@ func main() {
...
-		}()
-	}
-
-	// Parsing YAML config files
-	cfg, err := config.NewSystemProbeConfig(loggerName, opts.configPath)
-	if err != nil {
-		log.Criticalf("Failed to create agent config: %s", err)
-		os.Exit(1)
-	}
+	exit := make(chan bool)
```
Comment (sunhay) [inline cmd/system-probe/main.go]: This isn't being used atm. `runAgent()` is recreating this channel 	

```
e := make(chan bool)
go util.HandleSignals(e)
<-e
```

You can remove it as a parameter

File: pkg/ebpf/tracer.go [hunk 3]
```diff
@@ -113,6 +113,11 @@ func NewTracer(config *Config) (*Tracer, error) {
 		return nil, fmt.Errorf("could not load bpf module: %s", err)
 	}
 
+	// make sure debugfs is mounted
```
Comment (sunhay) [inline pkg/ebpf/tracer.go]: You've changed the order of operations here - this should be the very first thing we check when creating a `Tracer`.

File: pkg/ebpf/tracer_unsupported.go [hunk 4]
```diff
@@ -7,10 +7,9 @@ func CurrentKernelVersion() (uint32, error) {
 	return 0, ErrNotImplemented
 }
 
-// Tracer is not implemented on non-linux systems
 type Tracer struct{}
 
-// NewTracer is not implemented on non-linux systems
+// Tracer is not implemented on non-linux/Windows systems
```
Comment (sunhay) [inline pkg/ebpf/tracer_unsupported.go]: Just change this to `Tracer is not implemented`

File: pkg/ebpf/tracer_windows.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+package ebpf
+
+import (
+	"github.com/DataDog/datadog-agent/pkg/process/util"
+	"sync"
+)
+
+// CurrentKernelVersion is not implemented on non-linux systems
+func CurrentKernelVersion() (uint32, error) {
+	return 1, nil
```
Comment (sunhay) [inline pkg/ebpf/tracer_windows.go]: This should return a "not implemented" error

File: pkg/ebpf/tracer_windows.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+type Tracer struct {
+	config      *Config
+	portMapping *PortMapping
```
Comment (sunhay) [inline pkg/ebpf/tracer_windows.go]: You can remove all these fields outside of `Config`. They're not doing anything yet and we're not even sure which ones we want to keep.

File: pkg/ebpf/tracer_windows.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+	destExcludes   []*ConnectionFilter
+}
+
+func NewTracer(config *Config) (*Tracer, error) {
+	return &Tracer{}, nil
+}
+
+func (t *Tracer) Stop() {}
+
+func (t *Tracer) GetActiveConnections(_ string) (*Connections, error) {
```
Comment (sunhay) [inline pkg/ebpf/tracer_windows.go]: 👍

File: pkg/ebpf/windows_utils.go [hunk 8]
```diff
@@ -0,0 +1,8 @@
+// +build windows
+
+package ebpf
+
```
Comment (sunhay) [inline pkg/ebpf/windows_utils.go]: nit: extra whitespace line

File: pkg/ebpf/windows_utils.go [hunk 9]
```diff
@@ -0,0 +1,8 @@
...
+
+func verifyOSVersion(kernelCode uint32, platform string, exclusionList []string) (bool, string) {
+	return true, ""
```
Comment (sunhay) [inline pkg/ebpf/windows_utils.go]: Add a TODO that we this still needs to be fleshed out for the windows versions we support + that we should also probably remove `kernelCode` from it's parameters.

File: cmd/system-probe/probe.go [hunk 10]
```diff
@@ -42,28 +42,28 @@ func CreateSystemProbe(cfg *config.AgentConfig) (*SystemProbe, error) {
...
 
-	// Setting up the unix socket
-	uds, err := net.NewUDSListener(cfg)
+	var conn net.Conn
+	if runtime.GOOS != "windows" {
+		conn, err = net.NewUDSListener(cfg)
+	} else {
+		conn, err = net.NewListener(cfg)
+	}
+
```
Comment (sunhay) [inline cmd/system-probe/probe.go]: Not a fan of this - we can use build tags for this and have a simple `net.NewListener` that uses `NewUDSListener` for linux and something else for windows.

File: pkg/process/net/network_nolinux.go [hunk 11]
```diff
@@ -1,4 +1,4 @@
-// +build !linux
```
Comment (sunhay) [inline pkg/process/net/network_nolinux.go]: rename this file `network_unsupported.go`

File: pkg/ebpf/linux_utils.go [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+		return true, ""
+	}
+	errMsg := fmt.Sprintf("Kernel unsupported (%s) - ", kernelCodeToString(kernelCode))
+	errMsg += fmt.Sprintf("required functions missing: %s", strings.Join(missing, ", "))
+	return false, errMsg
+}
+
+// CurrentKernelVersion exposes calculated kernel version - exposed in LINUX_VERSION_CODE format
+// That is, for kernel "a.b.c", the version number will be (a<<16 + b<<8 + c)
+func CurrentKernelVersion() (uint32, error) {
```
Comment (sunhay) [inline pkg/ebpf/linux_utils.go]: Does this still need to be unexported

File: pkg/process/net/network_linux.go [hunk 13]
```diff
@@ -2,144 +2,9 @@
```
Comment (sunhay) [inline pkg/process/net/network_linux.go]: Lets call this `types_linux.go`

File: pkg/process/net/network_windows.go [hunk 14]
```diff
@@ -0,0 +1,10 @@
+// +build windows
```
Comment (sunhay) [inline pkg/process/net/network_windows.go]: `types_windows.go`

File: pkg/ebpf/linux_utils.go (hunk 12 above)
Comment (stb2132) [reply pkg/ebpf/linux_utils.go]: Nope! Changed it to not be exported.

Comment (stb2132) [reply pkg/ebpf/linux_utils.go]: Ah, actually used in nettop, so going to change this back to being exported for now.

File: pkg/ebpf/common.go [hunk 15]
```diff
@@ -56,7 +55,9 @@ func linuxKernelVersionCode(major, minor, patch uint32) uint32 {
 // along with some context on why it's not supported
 func IsTracerSupportedByOS(exclusionList []string) (bool, string) {
 	currentKernelCode, err := CurrentKernelVersion()
-	if err != nil {
+	if err == ErrNotImplemented {
```
Comment (stb2132) [inline pkg/ebpf/common.go]: This is to ensure that windows will build/not fail

File: pkg/ebpf/utils_unsupported.go [hunk 16]
```diff
@@ -0,0 +1,8 @@
+// +build !linux_bpf,!windows
+
+package ebpf
+
+// TODO Determine which windows versions we will support, and potentially remove kernelCode from parameters list
```
Comment (sunhay) [inline pkg/ebpf/utils_unsupported.go]: This comment doesn't make sense here - let's remove it

File: pkg/ebpf/utils_unsupported.go [hunk 17]
```diff
@@ -0,0 +1,8 @@
...
+func verifyOSVersion(kernelCode uint32, platform string, exclusionList []string) (bool, string) {
+	return true, ""
```
Comment (sunhay) [inline pkg/ebpf/utils_unsupported.go]: `return false, "unsupported architecture"`

File: pkg/process/net/types_unsupported.go [hunk 18]
```diff
@@ -1,4 +1,4 @@
-// +build !linux
+// +build !linux,!windows
```
Comment (sunhay) [inline pkg/process/net/types_unsupported.go]: nit: you need to update the comments in this file since they're all referencing linux specifically

**Review by sunhay — APPROVED** [review]

**Pull request merged**